### ISSUE TITLE
Support changing log level of apps while they're running from the dashboard

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/aspire#1193]
It would be super useful to be able to change the log level of an app (or all apps) from the dashboard while the apps are running. This would make it very easy to adjust logging level for diagnostics purposes while the app is running while looking at the log output in the dashboard, without needing to find the logging configuration of the app and change it manually, e.g. in the *appsettings.Development.json* file (which is persistent too).

While the JSON config source added by default to read config from *appsettings.json* and *appsettings.[EnvironmentName].json* already responds to file changes in real-time, it would be problematic for a few reasons to have the dashboard try to update this log-level in real-time, e.g. operations on the dashboard shouldn't have side effects on the source of the apps, how would it work in non-development scenarios when the dashboard is deployed and the apps are running in containers, etc.

## Strawman proposal

A new configuration provider could be added by the ServiceDefaults project that activates when an environment variable, e.g. `DOTNET_LOGGER_CONFIG_URL`, is populated with a URL of a remote HTTP endpoint that provides an API surface area for remotely providing logger configuration. The dashboard would set this environment variable when launching the app (similar to the OTLP endpoint).

The configuration provided by the service would take precedence over all other default added configuration sources (i.e. be added last), but furthermore, must support overriding the logger specific configuration from the *appsettings.json* files. This is because logger configuration is applied by prefix-based matching, in addition to a specified default, not just config provider ordering, and by default the logger level is set to "Information" for all sources in *appsettings.json*, and "Warning" for others like "Microsoft.AspNetCore".

Perhaps the app will need to post its current logger config to the service so that the dashboard can return the appropriately combined result, or perhaps the configuration provider added by ServiceDefaults will need to manually adjust the logger filter in addition to updating the configuration provider values.

If the endpoint must be protected from unauthorized access, some thought will be required as to how to facilitate this in a way that is durable across dashboard restarts when not run as part of the development scenario, as the lifetime of the apps may differ to that of the dashboard which makes session key style solutions impractical, e.g. a random key generated on dashboard start could not be injected into the already running apps if the apps are not started by the AppHost.

**Comment by bart-vmware** [comment]
> perhaps the configuration provider added by ServiceDefaults will need to manually adjust the logger filter in addition to updating the configuration provider values.

I've prototyped this approach for Steeltoe, but it isn't reliable due to the precedence rules inside LoggerFactory. I don't remember the exact details, but amongst other things it prioritizes longer keys. For example, let's say appsettings.json contains a level for "A.B" and "A.B.C.D.E.F", now the developer changes the level through the dashboard for A.B.C (expecting to change the level for all downstream keys). Just adding a filter rule for "A.B.C" will work for all downstream keys _except_ the existing "A.B.C.D.E.F", because it is more specific, so it overrules the level from "A.B.C".

**Comment by DamianEdwards** [comment]
@bart-vmware yeah that was the kind of thing I was thinking about that might mean we have to do more to support this well.

**Comment by leslierichardson95** [comment]
CC: @samsp-msft please help investigate :)
